spec/unit/puppetfile_spec.rb

```diff
@@ -68,34 +68,70 @@
-      path = File.join(PROJECT_ROOT, 'spec', 'fixtures', 'unit', 'puppetfile', 'valid-forge-with-version')
-      subject = described_class.new(path, {})
+    context 'using load' do
+      it "returns the loaded content" do
+        path = File.join(PROJECT_ROOT, 'spec', 'fixtures', 'unit', 'puppetfile', 'valid-forge-with-version')
+        subject = described_class.new(path, {})
 
-      loaded_content = subject.load
-      expect(loaded_content).to be_an_instance_of(Hash)
+        loaded_content = subject.load
+        expect(loaded_content).to be_an_instance_of(Hash)
 
-      has_some_data = loaded_content.values.none?(&:empty?)
-      expect(has_some_data).to be true
-    end
+        has_some_data = loaded_content.values.none?(&:empty?)
+        expect(has_some_data).to be true
+      end
 
-    it "is idempotent" do
-      path = File.join(PROJECT_ROOT, 'spec', 'fixtures', 'unit', 'puppetfile', 'valid-forge-with-version')
-      subject = described_class.new(path, {})
+      it "is idempotent" do
+        path = File.join(PROJECT_ROOT, 'spec', 'fixtures', 'unit', 'puppetfile', 'valid-forge-with-version')
+        subject = described_class.new(path, {})
 
-      expect(subject.loader).to receive(:load).and_call_original.once
+        expect(subject.loader).to receive(:load).and_call_original.once
 
-      loaded_content1 = subject.load
-      expect(subject.loaded?).to be true
-      loaded_content2 = subject.load
+        loaded_content1 = subject.load
+        expect(subject.loaded?).to be true
+        loaded_content2 = subject.load
 
-      expect(loaded_content2).to eq(loaded_content1)
+        expect(loaded_content2).to eq(loaded_content1)
+      end
+
+      it "returns nil if Puppetfile doesn't exist" do
+        path = '/rando/path/that/wont/exist'
+        subject = described_class.new(path, {})
+        expect(subject.load).to eq nil
+      end
     end
 
-    it "returns false if Puppetfile doesn't exist" do
-      path = '/rando/path/that/wont/exist'
-      subject = described_class.new(path, {})
-      expect(subject.load).to eq false
+    context 'using load!' do
+      it "returns the loaded content" do
+        path = File.join(PROJECT_ROOT, 'spec', 'fixtures', 'unit', 'puppetfile', 'valid-forge-with-version')
+        subject = described_class.new(path, {})
+
+        loaded_content = subject.load!
+        expect(loaded_content).to be_an_instance_of(Hash)
+
+        has_some_data = loaded_content.values.none?(&:empty?)
+        expect(has_some_data).to be true
+      end
+
+      it "is not idempotent" do
+        path = File.join(PROJECT_ROOT, 'spec', 'fixtures', 'unit', 'puppetfile', 'valid-forge-with-version')
+        subject = described_class.new(path, {})
+
+        expect(subject.loader).to receive(:load).and_call_original.twice
+
+        loaded_content1 = subject.load!
+        expect(subject.loaded?).to be true
+        expect {
+          subject.load!
+        }.to raise_error(R10K::Error)
```

Comment:
Why does this error? Is there a specific error message we should be looking for?

Reply:
It raises saying that duplicate modules are not allowed, however I don't know that that error is important so long as it does error with an nice exception. Here I use R10K::Error as a stand in for a "nice" error since that means the error will be caught and wrapped with appropriate logging. I can add a comment to that effect?

Reply:
added something to that effect

Reply:
hmm okay. I guess I'm not really understanding why it's important to ensure that `load!` is not idempotent at all..

Reply:
that's fair, was I really just mirroring the existing tests.

Reply:
removed